### ISSUE TITLE
T cell classification as example dataset

### ISSUE THREAD

**Issue by agitter** [gitter-lab/ml-bio-workshop#51]
@xiaohk has been working with Melissa Skala's group on a T cell image classification project.  It is related to the [Walsh et al. 2019 preprint](https://doi.org/10.1101/536813) but uses a different type of image.

This could be a great new motivating example and dataset for the workshop.  It's a setting where machine learning is clearly needed because to be practically useful, cells would need to be classified extremely quickly.  There is not enough time for manual expert labeling.  Also, Jay has simplified versions of the dataset where we have only two features, which we could use for interactive activities on paper and in the software.

**Comment by xiaohk** [comment]
Here is the background of this dataset and study.

# Dataset description

This microscopy dataset includes gray scale images of two type of T cells: activated and quiescent (not activated). These T cells come from 6 donors.

|Activated|Quiescent|
|:---:|:---:|
|![CD3_2_act_2_6_66](https://user-images.githubusercontent.com/15007159/61666368-e4804d00-ac9c-11e9-9031-a3f9f6cfd7b1.png)|![CD8_2_noact_3_3_13](https://user-images.githubusercontent.com/15007159/61666346-d9c5b800-ac9c-11e9-9044-e13c218d0da0.png)|

For our T cell study, we sub-sampled 10% images from the full dataset, and used them to demonstrate our workflow through [Jupyter notebooks](https://github.com/gitter-lab/t-cell-classification). Since WARF team is filing a patent for this study, we can only make these subsampled images public.

||Donor 1|Donor 2|Donor 3|Donor 4|Donor 5|Donor 6|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Full Activated|271|656|487|494|694|446|
|Full Quiescent|1631|152|1276|1629|252|589|
|Subsampled Activated|27|65|48|49|69|44|
|Subsampled Quiescent|163|15|127|162|25|58|

## Machine Learning Methods

The goal of this study is to develop a method to classify T cell activation stage (activated vs. quiescent). We have explored many classifiers.

| Model                                             | Description                                                                                                                                                                                      |
|---------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Frequency Classifier                              | Predict class probability using the class frequencies in the training set.                                                                                                                       |
| Logistic Regression with Pixel Intensity          | Regularized logistic regression model fitted with the image pixel intensity matrix (82 x 82). Regularization power λ of L1 penalty is tuned.                   |
| Logistic Regression with Total Intensity and Size | Regularized logistic regression model fitted with two numerical values: image total intensity and cell mask size. Regularization power λ of L1 penalty is tuned.                   |
| Logistic Regression with CellProfiler Features    | Regularized logistic regression model fitted with 123 features extracted from CellProfiler related to intensity, texture, and area. Regularization power λ of L1 penalty is tuned. |
| One-layer Fully Connected Neural Network          | Fully connected one-hidden-layer neural network with pixel intensity as input. Number of neurons, learning rate, and batch size are tuned.                                                       |
| LeNet CNN                                         | CNN with the LeNet architecture with pixel intensity as input. Learning rate and batch size are tuned.                                                                                           |
| Pre-trained CNN Off-the-shelf Model               | Freeze layers of a pre-trained Inception v3 CNN. Train a final added layer from scratch with extracted off-the-shelf features. Learning rate and batch size are tuned.                           |
| Pre-trained CNN with Fine-tuning                  | Fine-tune the last n layers of a pre-trained Inception v3 CNN. The layer number n, learning rate, and batch size are tuned.                                                                  |

Here is the test result of these classifiers:

![metric](https://user-images.githubusercontent.com/15007159/61667809-9ff6b080-aca0-11e9-9f11-875f49b94344.png)

## ML4BIO Ideas

1. Introduce basic image processing literacy
    - Cropping, padding (adding black borders to make all images have the same size)
    - Image augmentation (rotation, mirroring)
    - Image thresholding (useful for segmentation, not used in our study though)
    - Image entropy (useful for filtering out poorly cropped images)
2. Simple classifiers
    - Logistic regression + two features (total intensity and cell mask size)
    - Logistic regression + image pixel matrix (82 x 82 features)
    - Logistic regression + CellProfiler features (we extracted 123 features, but audience might not be familiar with this software/pipeline)
3. Unbalanced dataset
    - Use class weighting during training
    - Prefer average precision over ROC AUC or accuracy
4. Classifier interpretation
    - Interpret logistic regression coefficients
    - Use UMAP to visualize different features used in logistic regression models
5. Cross-validation
   - We designed a complex and interesting nested cross-validation scheme
   - It iterates through donors, so validation set and test set always have a different donor from the training set
6. More...

|![cross_validation](https://user-images.githubusercontent.com/15007159/61668549-9ec68300-aca2-11e9-8d90-6da4412e0a0b.png)|
|:---:|
|Our nested cross-validation demo|

**Comment by cmilica** [comment]
This is great - I think the simplified version is what we want to focus on. 

These ideas are great! @agitter maybe we can go through them in our next meeting? 

One thing that would be very important, since we will be using this as a "motivation" example for "Why bother to use Machine Learning in the first place", @xiaohk can you briefly sum up:

- how was Ml beneficial vs. using some other approach? 
- why is the detection of active/inactive T-cells important and what major contribution did it make? 

Thanks!

**Comment by xiaohk** [comment]
- People want to classify T cells quickly. Using a sophisticated machine learning classifier is much more effective than labeling T cells manually.
- Immunotherapy is a type of cancer treatment that uses the body’s own immune cells to boost natural defenses against cancer, and T cells are a popular target for immunotherapies. To fully optimize the effect of immunotherapy, T cell activation must be assessed at a single-cell level. Melissa Skala's group developed a label-free and non-destructive autofluorescence imaging method to easily and quickly acquire T cell intensity images. Therefore, an activated/quiescent (inactive) T cell classifier that uses autofluorescence intensity images can contribute to the applications of immunotherapy.

Let me know if you have other questions 😄

**Comment by cmilica** [comment]
@xiaohk thank you!

**Comment by agitter** [comment]
Hopefully we'll be able to publicly post our manuscript as a preprint soon, which will further explain the motivation and contribution.

Here are some example slides I've used when using this project to introduce classification:
[t-cell-slides.pptx](https://github.com/gitter-lab/ml-bio-workshop/files/3448247/t-cell-slides.pptx)

The slides also help explain why ML is needed.  The most advanced classifiers are extremely accurate even though the predictive task is hard for a non-expert human.  The grayscale images can be quite blurry and hard to distinguish.  As @xiaohk said, however, speed is the dominant reason we need ML.

@xiaohk what do you think about preparing simple csv files with the features and class labels for the 10% subsampled images?  The ml4bio software uses a csv format with feature columns first and the last column for the class, as in [this example](https://github.com/gitter-lab/ml-bio-workshop/blob/master/data/prostate/prostate.csv).  We could provide the raw pixels, size + intensity, and CellProfiler features as separate versions of the data if we want to discuss the importance of using the right features in the workshop.

**Comment by xiaohk** [comment]
https://github.com/gitter-lab/ml-bio-workshop/pull/55 adds size + intensity and CellProfiler features CSV files. Since the CSV files for the raw pixels are two large (60MB for augmented images, and 10MB for unaugmented images), we will decide where to store them later.

Here are the descriptions for all CSV files:

|File|Description|
|:---:|:---:|
|`pixel_feature_augmented.csv`|Raw pixel intensity of images. Each cell image has dimension 78x78, so there are 6084 raw pixel features for each row. This file includes features extracted from **augmented** images (3 rotated images and 2 mirrored images).|
|`pixel_feature_unaugmented.csv`|Raw pixel intensity of images. Each cell image has dimension 78x78, so there are 6084 raw pixel features for each row. This file only includes features extracted from **unaugmented** images.|
|`size_intensity_feature.csv`|This file includes two features extracted from unaugmented images: cell mask size and cell image total intensity.|
|`cellprofiler_feature.csv`|This file includes 123 features extracted from unaugmented images using the software CellProfiler. These 123 features covers pipelines `MeaureObjectSizeShape`, `MeasureObjectIntensity`, and `MeasureTexture`.|

- The row indices of `pixel_feature_unaugmented.csv`, `size_intensity_feature.csv`, and `cellprofiler_feature.csv` are the same.

**Comment by agitter** [comment]
Thanks for adding the datasets in #55.  We can decide later how we want to handle the donor ID for the workshop.  Currently, our software would treat the donor ID as a feature.

Should we add `pixel_feature_augmented.csv` and `pixel_feature_unaugmented.csv` to the `resource` subdirectory of https://github.com/gitter-lab/t-cell-classification?  Or add these to your Zenodo dataset?  We wanted to keep the size of this repository small because each workshop participant has to download it, but the file sizes may be small enough for https://github.com/gitter-lab/t-cell-classification and not require Zenodo.

**Comment by xiaohk** [comment]
Sounds good. https://github.com/gitter-lab/t-cell-classification/commit/535d477169a306a8bfd7a18893d0ea721649c5b5 adds these two csv files to https://github.com/gitter-lab/t-cell-classification. I renamed them to `pixel_features_augmented.csv` and `pixel_features_unaugmented.csv` to be consistent with other files in the `resource` directory.
